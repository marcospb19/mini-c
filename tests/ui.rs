//! This is a test module that creates snapshots, and might do
//! UI testing in the future.
//!
//! This shouldn't be called ui.rs hmmmmmmmm.
//!
//! Snapshot todo list:
//! - [X] Source code.
//! - [X] AST.
//! - [ ] LLVM IR.
//! - [ ] Execution stdout output (UI).
//! - [ ] (maybe) Execution final state (variables, functions).

use fs_err as fs;
use mini_c::test_utils;

macro_rules! snapshot_example_file {
    ($example_path:literal) => {
        let file_contents = &fs::read_to_string($example_path)
            .unwrap_or_else(|err| panic!("Failed to read file {}: {err}.", $example_path));
        // Source code
        insta::assert_display_snapshot!(file_contents);
        // AST
        insta::assert_ron_snapshot!(test_utils::generate_ast(file_contents));
    };
}

#[test]
fn example_break() {
    snapshot_example_file!("examples/break.c");
}
#[test]
fn example_coins() {
    snapshot_example_file!("examples/coins.c");
}
#[test]
fn example_conds() {
    snapshot_example_file!("examples/conds.c");
}
#[test]
fn example_easy() {
    snapshot_example_file!("examples/easy.c");
}
#[test]
fn example_fib() {
    snapshot_example_file!("examples/fib.c");
}
#[test]
fn example_file_1() {
    snapshot_example_file!("examples/file_1.c");
}
#[test]
fn example_file_2() {
    snapshot_example_file!("examples/file_2.c");
}
#[test]
fn example_if_pass() {
    snapshot_example_file!("examples/if-pass.c");
}
#[test]
fn example_operators() {
    snapshot_example_file!("examples/operators.c");
}
#[test]
fn example_queen() {
    snapshot_example_file!("examples/queen.c");
}
#[test]
fn example_var() {
    snapshot_example_file!("examples/var.c");
}
#[test]
fn example_var_loop() {
    snapshot_example_file!("examples/var_loop.c");
}
#[test]
fn example_while() {
    snapshot_example_file!("examples/while.c");
}
