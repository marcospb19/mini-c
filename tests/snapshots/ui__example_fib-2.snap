---
source: tests/ui.rs
expression: test_mini_c(file_contents)
---
Program(
  preamble: true,
  declarations: [
    Variable(VariableDeclaration(Int, [
      Normal("i"),
      Normal("result"),
    ])),
    Function(FunctionDeclaration(Int, "fib", [
      Parameter(Int, "t"),
    ], Some(Scope([], [
      If(IfStatement(Binary(BinaryExpression(VariableReference(Normal("t")), Less, Value(Int(1)))), Scope([], [
        Return(ReturnStatement(Some(Value(Int(0))))),
      ]), None)),
      If(IfStatement(Binary(BinaryExpression(VariableReference(Normal("t")), Equals, Value(Int(1)))), Scope([], [
        Return(ReturnStatement(Some(Value(Int(1))))),
      ]), None)),
      Return(ReturnStatement(Some(Binary(BinaryExpression(FunctionCall(FunctionCallExpression("fib", [
        Binary(BinaryExpression(VariableReference(Normal("t")), Sub, Value(Int(1)))),
      ])), Add, FunctionCall(FunctionCallExpression("fib", [
        Binary(BinaryExpression(VariableReference(Normal("t")), Sub, Value(Int(2)))),
      ]))))))),
    ])))),
    Function(FunctionDeclaration(Int, "main", [], Some(Scope([], [
      Expression(Assignment(AssignmentExpression(Normal("i"), Value(Int(10))))),
      Expression(Assignment(AssignmentExpression(Normal("result"), FunctionCall(FunctionCallExpression("fib", [
        VariableReference(Normal("i")),
      ]))))),
      Expression(FunctionCall(FunctionCallExpression("putint", [
        VariableReference(Normal("result")),
      ]))),
      Return(ReturnStatement(Some(Value(Int(0))))),
    ])))),
  ],
)
